docs/microbenchmark-design-guidelines.md

```diff
@@ -0,0 +1,557 @@
+
+* No, because it's an edge case.
+* No, because the same code path is executed for big buffers (`len > 64 && len > MemmoveNativeThreshold`)
+
+What should we test? This method is commonly used, so we should test all four code paths:
+
+* small buffers (`len <= 16`)
+* medium size buffers (`len > 16 && len <= 64`)
+* big buffers (`len > 64 && len < MemmoveNativeThreshold`)
+* very big buffers (`len > 64 && len > MemmoveNativeThreshold`)
+
+One important thing is to leave a comment in the source code and explain our decision. So nobody removes an important test case in the future and we don't forget why we added it ourselves. An example:
+
+```cs
+[Params(
+    128, // stackalloc path
+    1024 * 256)] // ArrayPool.Shared.Rent without allocation 
+public int Count { get; set; }
+```
+
+### Always the same input data
+
+To allow for apples to apples comparison over time, the input data should be always the same.
+
+`ValuesGenerator` class from `BenchmarkDotNet.Extensions` library exposes following methods that guarantee to always produce the same data:
+
+```cs
+T GetNonDefaultValue<T>();
+T[] ArrayOfUniqueValues<T>(int count);
+T[] Array<T>(int count);
+Dictionary<TKey, TValue> Dictionary<TKey, TValue>(int count)
+```
+
+As of today, the `T` can be: `byte`, `char`, `int`, `double`, `bool` and `string`. Extending `T` to more types is very welcomed!
+
+**Note:** `ValuesGenerator` is simply always creating a new instance of `Random` with a constant seed. It's a crucial component and its correctness is verified using  [Unit Tests](https://github.com/dotnet/performance/blob/master/src/tests/harness/BenchmarkDotNet.Extensions.Tests/UniqueValuesGeneratorTests.cs).
+
+**Note:** Please don't use `Random` directly in the benchmarks, do use `ValuesGenerator` and extend it with missing features when needed.
+
+### BenchmarkDotNet
+
+#### Arguments
+
+`[Arguments]` is xUnit's `[InlineData]` counterpart.
+
+```cs
+[Benchmark]
+[Arguments(100)]
+public void Method(int length)
+```
+
+**Note:** BenchmarkDotNet supports passing arguments by `ref` to the benchmarked method.
+
+#### Params
+
+If you need to use the argument in the setup method, then instead of using `Arguments` you should use `Params`. A `[Params]` field|property can be accessed during the setup/cleanup ([docs](https://benchmarkdotnet.org/Advanced/Params.htm)). Example:
+
+```cs
+int[] _array;
+
+[Params(
+    100, // comment explaining why it's needed to have this test case
+    1_000, // comment explaining why it's needed to have this test case
+    10_000)] // comment explaining why it's needed to have this test case
+public int Size { get; set; }
+
+[GlobalSetup]
+public void PrepareArray() => _array = new int[Size];
+```
+
+However, **`[Params]` are applied to all the benchmarks in given class**. So if few benchmarks require different `[Params]` values they have to be split into separate types (current BenchmarkDotNet implementation limitation).
```

Comment:
Is there a BDN issue tracking this limitation? Will it go away in the future? If not, I'd elide the parenthetical here and leave the statement as is.

Reply:
I made the following sentence bold and underline. + "NOT" as uppercase (Vance style ;) )

<ins>**If using `[GlobalSetup]` is enough, you should NOT be using `[IterationSetup]`**</ins>